Source/WebCore/css/CSSCounterStyle.cpp

```diff
@@ -0,0 +1,287 @@
+            m_extendsName = secondValue && secondValue->isCustomIdent() ? makeAtomString(secondValue->cssText()) : makeAtomString("decimal"_s);
+        } else if (system() == CounterStyleSystem::Fixed) {
+            ASSERT(secondValue && secondValue->isInteger());
+            // If no value is provided after `fixed`, the first synbol value is implicitly 1 (https://www.w3.org/TR/css-counter-styles-3/#first-symbol-value).
+            m_fixedSystemFirstSymbolValue = secondValue && secondValue->isInteger() ? secondValue->intValue() : 1;
+        }
+    }
+}
+
+CounterStyle::CounterStyle(Name name, CounterStyleSystem system, NegativeSymbols negativeSymbols, Symbol prefix, Symbol suffix, Ranges&& ranges, Pad pad, Name fallback, Vector<Symbol>&& symbols, CounterStyleSpeakAs speakAs, Name speakAsReference)
+: m_predefinedCounterStyle { true }, m_name { name }, m_system { system }, m_negativeSymbols { negativeSymbols }, m_prefix { prefix }, m_suffix { suffix }, m_ranges { WTFMove(ranges) }, m_pad { pad }, m_fallbackName { fallback }, m_symbols { WTFMove(symbols) }, m_speakAs { speakAs }, m_speakAsNameReference { speakAsReference }
+{ }
+
+CounterStyle::CounterStyle(const StyleRuleCounterStyle& rule)
+: m_name { rule.name() }, m_system { toCounterStyleSystemEnum(rule.systemCSSValue()) }, m_negativeSymbols { translateNegativeSymbolsFromStyleRule(rule) }, m_prefix { translatePrefixFromStyleRule(rule) }, m_suffix { translateSuffixFromStyleRule(rule) }, m_ranges { translateRangeFromStyleRule(rule) }, m_pad { translatePadFromStyleRule(rule) }, m_fallbackName { translateFallbackNameFromStyleRule(rule) }, m_symbols { translateSymbolsFromStyleRule(rule) }, m_speakAs { CounterStyleSpeakAs::Auto }, m_speakAsNameReference { ""_s }
+{
+    extractDataFromSystemDescriptor(rule);
+    // Marking explicitly defined descriptors such that Extends system knows which properties to extend.
+    if (rule.systemCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::System);
+    if (rule.negativeCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::Negative);
+    if (rule.prefixCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::Prefix);
+    if (rule.suffixCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::Suffix);
+    if (rule.rangeCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::Range);
+    if (rule.padCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::Pad);
+    if (rule.fallbackCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::Fallback);
+    if (rule.symbolsCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::Symbols);
+    if (rule.additiveSymbolsCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::AdditiveSymbols);
+    if (rule.speakAsCSSValue())
+        m_explicitlySetDescriptors.add(ExplicitlySetDescriptors::SpeakAs);
+}
```

Comment:
It might be better to not trying to implement all descriptors in one go. Maybe just start with a few of the most common ones and add the rest when the whole mechanism is in place? It would make these patches easier to follow and review.